### PR TITLE
Added JSON support

### LINKED ISSUE #1234: Add Schemaless/ JSON support.
We want support for schemaless. 

The plan is to rely on tantivy's JSON field, in two ways:
- [x] json type.
- [ ] a catchall json field that captures all of the unmatched fields, if dynamic is set to true.

### PR DESCRIPTION
**Pull request by fulmicoton**
Dynamic mapping in quickwit.
    
    - Adds the JSON field type. This feature relies heavily on tantivy#1270.
    - More unit test for deserialization of the doc mapper.
    - Introduce the notion of mode in the default_mapper to decide how
      unmapped fields should be handled. (lenient, strict, dynamic)
      - In dynamic mode, unmapped fields are stored in a _dynamic field.
      - The DefaultDocMapper is refactored, yielding a substantial
      performance improvement in indexing when the data contains many short
      or untokenized fields. (Closes Document parsing is needlessly slow. #1290)
      - field_mapping can be absent of the index conf. (Closes Allow missing field mappings in conf #1324)
      - Error on deserializaton for unknown field in the default doc mapper.
      - When an error is spotted on doc parsing, the error contains the entire field path from the root instead of just the
      faulty leaf field name.
    
    Closes #1234
    
    Json output will be addressed in a separate ticket. #1325

### CODE REVIEW

File: quickwit-config/src/index_config.rs [hunk 1]
```diff
@@ -306,17 +311,21 @@ pub fn build_doc_mapper(
     search_settings: &SearchSettings,
     indexing_settings: &IndexingSettings,
 ) -> anyhow::Result<Arc<dyn DocMapper>> {
-    let mut builder = DefaultDocMapperBuilder::new();
```
Comment (fulmicoton) [inline quickwit-config/src/index_config.rs:309]: That change may seem weird... 

The way we build DefaultDocMapper is actually quite unexpected.
We both build it through deserializing and through a `DocMapping` object.

For this reason it is quite easy to forget modify both location if we add a field for instance.

Relying on struct initialization makes this kind of mistakes  harder.

Comment (fmassot) [reply quickwit-config/src/index_config.rs:309]: Thanks for cleaning that, I think it comes from the 0.2 hurry release.

File: quickwit-doc-mapper/src/default_doc_mapper/default_mapper.rs [hunk 2]
```diff
@@ -369,42 +83,39 @@ impl From<DefaultDocMapper> for DefaultDocMapperBuilder {
 #[derive(Serialize, Deserialize, Clone)]
 #[serde(try_from = "DefaultDocMapperBuilder", into = "DefaultDocMapperBuilder")]
 pub struct DefaultDocMapper {
-    /// Store the json source in a text field _source.
-    pub store_source: bool,
+    /// Field in which the source should be stored.
+    pub(crate) source_field: Option<Field>,
```
Comment (fmassot) [inline quickwit-doc-mapper/src/default_doc_mapper/default_mapper.rs:87]: I was a bit surprised to see a tantivy `Field` as usually we don't want to store it directly, it's only relevant to the `DefaultDocMapper` schema and not for an external schema. I would avoid having putting a `pub` here and enrich a bit the docs `/// Field in which the source should be stored, can only be used with a `DefaultDocMapper` schema.`

File: quickwit-doc-mapper/src/default_doc_mapper/mod.rs [hunk 3]
```diff
@@ -18,28 +18,32 @@
...
 
 /// Validates a field mapping name.
 /// Returns `Ok(())` if the name can be used for a field mapping. Does not check for reserved field
 /// mapping names such as `_source`.
 ///
 /// A field mapping name:
 /// - may only contain uppercase and lowercase ASCII letters `[a-zA-Z]`, digits `[0-9]`, hyphens
-///   `-`, periods `.`, and underscores `_`;
-/// - must start with an uppercase or lowercase ASCII letter `[a-zA-Z]`, or an underscore `_`;
+///   `-`, and underscores `_`;
```
Comment (fmassot) [inline quickwit-doc-mapper/src/default_doc_mapper/mod.rs:44]: I'm ok to ban field names that start with `_`. I'm expecting to see users with fields starting `_` to be happy with json field.

File: quickwit-doc-mapper/src/doc_mapper.rs [hunk 4]
```diff
@@ -133,4 +137,103 @@ mod tests {
...
+            sort_by_field: None,
+            aggregation_request: None,
+        };
+        let query = doc_mapper.query(schema, &search_request).unwrap();
+        assert_eq!(
+            format!("{:?}", query),
+            r#"BooleanQuery { subqueries: [(Should, TermQuery(Term(type=Json, field=0, path=toto, vtype=U64, 5))), (Should, TermQuery(Term(type=Json, field=0, path=toto, vtype=Str, "5")))] }"#
+        );
+    }
 }
```
Comment (fmassot) [inline quickwit-doc-mapper/src/doc_mapper.rs:239]: Currently, we can query a text field `body` with the query `body.wrong_subpath:hello`. 

Here is the test that is passing:

```rust
#[test]
    fn test_doc_mapper_accept_sub_field_query_on_non_json_field() {
        let mut doc_mapper_builder = DefaultDocMapperBuilder::default();
        doc_mapper_builder.field_mappings.push(FieldMappingEntry {
            name: "json_field".to_string(),
            mapping_type: FieldMappingType::Json(
                QuickwitJsonOptions::default(),
                Cardinality::SingleValue,
            ),
        });
        doc_mapper_builder.field_mappings.push(FieldMappingEntry {
            name: "body".to_string(),
            mapping_type: FieldMappingType::Text(
                QuickwitTextOptions::default(),
                Cardinality::SingleValue,
            ),
        });
        
        let doc_mapper = doc_mapper_builder.build().unwrap();
        let schema = doc_mapper.schema();
        let search_request = SearchRequest {
            index_id: "quickwit-index".to_string(),
            query: "body.wrong_field:hello".to_string(),
            search_fields: vec![],
            start_timestamp: None,
            end_timestamp: None,
            max_hits: 10,
            start_offset: 0,
            sort_order: None,
            sort_by_field: None,
            aggregation_request: None,
        };
        let query = doc_mapper.query(schema, &search_request).unwrap();
        assert_eq!(
            format!("{:?}", query),
            r#"TermQuery(Term(type=Json, field=1, vtype=Str, "hello"))"#
        );
    }
```

File: quickwit-doc-mapper/src/error.rs [hunk 5]
```diff
@@ -17,6 +17,7 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
+use tantivy::schema::DocParsingError as TantivyDocParsingError;
 use thiserror::Error;
 
 // TODO improve me and my error messages :)
```
Comment (fmassot) [inline quickwit-doc-mapper/src/error.rs:23]: Either ditch this old comment or let's create an issue so that we can track it.

Comment (fmassot) [reply quickwit-doc-mapper/src/error.rs:23]: (I will create the issue related to query parsing error I think)

**Review by fmassot — APPROVED** [review]
Great work. I saw some typos in the documentation, I will add a commit to yours in a couple of hours or if it's already merged I will open a PR on that.

File: quickwit-doc-mapper/src/default_doc_mapper/default_mapper.rs (hunk 2 above)
Comment (fulmicoton) [reply quickwit-doc-mapper/src/default_doc_mapper/default_mapper.rs]: Good point.
I added the comment and made everything private.

I also opened #1340 to see if we can simplify the weird DocMapper contract.

**Pull request merged**